cmake/modules/FindFFMPEG.cmake

```diff
@@ -165,153 +265,49 @@ if(NOT ENABLE_INTERNAL_FFMPEG OR KODI_DEPENDSBUILD)
                                                     FFMPEG_VERSION
                                       FAIL_MESSAGE "FFmpeg ${REQUIRED_FFMPEG_VERSION} not found, please consider using -DENABLE_INTERNAL_FFMPEG=ON")
 
+    if(FFMPEG_FOUND)
+      set(FFMPEG_LDFLAGS ${PC_FFMPEG_LDFLAGS} CACHE STRING "ffmpeg linker flags")
+
+      set(FFMPEG_LIBRARIES ${FFMPEG_LIBAVCODEC} ${FFMPEG_LIBAVFILTER}
+                           ${FFMPEG_LIBAVFORMAT} ${FFMPEG_LIBAVUTIL}
+                           ${FFMPEG_LIBSWSCALE} ${FFMPEG_LIBSWRESAMPLE}
+                           ${FFMPEG_LIBPOSTPROC} ${FFMPEG_LDFLAGS})
+
+      # check if ffmpeg libs are statically linked
+      set(FFMPEG_LIB_TYPE SHARED)
+      foreach(_fflib IN LISTS FFMPEG_LIBRARIES)
+        if(${_fflib} MATCHES ".+\.a$" AND PC_FFMPEG_STATIC_LDFLAGS)
+          set(FFMPEG_LDFLAGS ${PC_FFMPEG_STATIC_LDFLAGS} CACHE STRING "ffmpeg linker flags" FORCE)
+          set(FFMPEG_LIB_TYPE STATIC)
+          break()
+        endif()
+      endforeach()
+    endif()
   else()
-    message(STATUS "FFmpeg ${REQUIRED_FFMPEG_VERSION} not found, falling back to internal build")
-    unset(FFMPEG_INCLUDE_DIRS)
-    unset(FFMPEG_INCLUDE_DIRS CACHE)
-    unset(FFMPEG_LIBRARIES)
-    unset(FFMPEG_LIBRARIES CACHE)
-    unset(FFMPEG_DEFINITIONS)
-    unset(FFMPEG_DEFINITIONS CACHE)
-  endif()
-
-  if(FFMPEG_FOUND)
-    set(FFMPEG_LDFLAGS ${PC_FFMPEG_LDFLAGS} CACHE STRING "ffmpeg linker flags")
-
-    # check if ffmpeg libs are statically linked
-    set(FFMPEG_LIB_TYPE SHARED)
-    foreach(_fflib IN LISTS FFMPEG_LIBRARIES)
-      if(${_fflib} MATCHES ".+\.a$" AND PC_FFMPEG_STATIC_LDFLAGS)
-        set(FFMPEG_LDFLAGS ${PC_FFMPEG_STATIC_LDFLAGS} CACHE STRING "ffmpeg linker flags" FORCE)
-        set(FFMPEG_LIB_TYPE STATIC)
-        break()
-      endif()
-    endforeach()
-
-    set(FFMPEG_LIBRARIES ${FFMPEG_LIBAVCODEC} ${FFMPEG_LIBAVFILTER}
-                         ${FFMPEG_LIBAVFORMAT} ${FFMPEG_LIBAVUTIL}
-                         ${FFMPEG_LIBSWSCALE} ${FFMPEG_LIBSWRESAMPLE}
-                         ${FFMPEG_LIBPOSTPROC} ${FFMPEG_LDFLAGS})
-    list(APPEND FFMPEG_DEFINITIONS -DFFMPEG_VER_SHA=\"${FFMPEG_VERSION}\")
-
-    if(NOT TARGET ffmpeg)
-      add_library(ffmpeg ${FFMPEG_LIB_TYPE} IMPORTED)
-      set_target_properties(ffmpeg PROPERTIES
-                                   FOLDER "External Projects"
-                                   IMPORTED_LOCATION "${FFMPEG_LIBRARIES}"
-                                   INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}"
-                                   INTERFACE_LINK_LIBRARIES "${FFMPEG_LDFLAGS}"
-                                   INTERFACE_COMPILE_DEFINITIONS "${FFMPEG_DEFINITIONS}")
+    if(FFMPEG_PATH)
```

Comment:
I guess we should also fail when ENABLE_INTERNAL_FFMPEG=OFF and no suitable ffmpeg was found.

Reply:
Just trying to get my head around what that would change.

ENABLE_INTERNAL_FFMPEG is set off when WITH_FFMPEG or FFMPEG_PATH is set, so essentially changing this if(FFMPEG_PATH) to if(NOT ENABLE_INTERNAL_FFMPEG)

My concern is, in what case is the fallthrough triggered then?

I dont think it would be possible for any fallthrough to work.

case 1: WITH_FFMPEG=... -> no fallthrough, FFMPEG_PATH is set line 143, ENABLE_INTERNAL_FFMPEG is set to OFF line 161
case 2: FFMPEG_PATH=... -> no fallthrough, ENABLE_INTERNAL_FFMPEG is set to OFF line 161
case 3: ENABLE_INTERNAL_FFMPEG=ON -> doesnt do this path
case 4: User provides none of the above -> no fallthrough, if(NOT ENABLE_INTERNAL_FFMPEG) will fail due to not being set

In the current state, fall through would occur in case 4, as i guess we assume if a user is setting WITH_FFMPEG or FFMPEG_PATH they dont want to do internal, they want to use that specific version given, so its ok/expected not to fallthrough and build internal

I guess the question is, is there ever any expectation to provide fallthrough in case 1 or 2, and also, are there any cases i havent provided that might give a different outcome?

Reply:
ad 1 and 2): no fallback

ad 4) this should work: https://github.com/wsnipex/xbmc/commit/af472688acb730a845a97f713fe6bde087b80e8e
thoughts?

Reply:
It does cover that.
```
if (internal)
else
  findpackages
  if(pkgconfig vars found)
  else
    if FFMPEG_PATH - (covers with_ffmpeg and/or ffmpeg_path)
      fail suggest enable_internal_ffmpeg
    else
      build internal fallback
    endif
  endif
endif 
```

so essentially, we dont care what state ENABLE_INTERNAL_FFMPEG is in (unset or off), if the pkgconfig fails to find the system, it'll build internal as long as with_ffmpeg or ffmpeg_path werent provided

Reply:
> so essentially, we dont care what state ENABLE_INTERNAL_FFMPEG is in (unset or off), if the pkgconfig fails to find the system, it'll build internal

that's exactly my point. I though it would be expected behavior to not fallback when the user expicitly turned it off with -DENABLE_INTERNAL_FFMPEG=OFF

anyway, it's not so important to stall the PR. We can always fix it, if someone complains.

Reply:
Ah k, i see what you are asking now. At that point i wonder why even bother with the fallthrough? Maybe its just better to error out and say, look, either pass the correct ffmpeg locations, or use internal, try again.